commit 7e8210156baa8fa2c859ace071e0a318710c5ec2
Author: Relic1212 <67271773+Relic1212@users.noreply.github.com>
Date:   Fri Nov 7 13:00:08 2025 +0100

    Revert "Only build KStartupInfo for X11"
    
    This reverts commit 2bc2cf7c45ceef4f44b5d5d5edb71cd7d95ad08d.

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index ba748c2..7fee002 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -16,6 +16,9 @@ set_target_properties(KF6WindowSystem PROPERTIES
 ecm_create_qm_loader(KF6WindowSystem kwindowsystem6_qt)
 
 target_sources(KF6WindowSystem PRIVATE
+    kkeyserver.cpp
+    kstartupinfo.cpp
+    kusertimestamp.cpp
     kwindoweffects.cpp
     kwindoweffects_dummy.cpp
     kwindowshadow.cpp
@@ -100,6 +103,9 @@ target_link_libraries(KF6WindowSystem
 
 ecm_generate_headers(KWindowSystem_HEADERS
   HEADER_NAMES
+  KKeyServer
+  KStartupInfo
+  KUserTimestamp
   KWindowEffects
   KWindowShadow,KWindowShadowTile
   KWindowSystem
diff --git a/src/kstartupinfo.cpp b/src/kstartupinfo.cpp
index 3038e62..d408381 100644
--- a/src/kstartupinfo.cpp
+++ b/src/kstartupinfo.cpp
@@ -21,30 +21,43 @@
 
 #include <QDateTime>
 
+#include <config-kwindowsystem.h> // KWINDOWSYSTEM_HAVE_X11
+
 // need to resolve INT32(qglobal.h)<>INT32(Xlibint.h) conflict
 #ifndef QT_CLEAN_NAMESPACE
 #define QT_CLEAN_NAMESPACE
 #endif
 
-#include <QTimer>
-#include <netwm.h>
-#include <stdlib.h>
+#ifndef Q_OS_WIN
 #include <sys/time.h>
 #include <unistd.h>
+#else
+#include <process.h>
+#include <winsock2.h>
+#endif
+#include <QTimer>
+#include <stdlib.h>
+#if KWINDOWSYSTEM_HAVE_X11
+#include <netwm.h>
+
+#include <private/qtx11extras_p.h>
 
+#endif
 #include <QCoreApplication>
 #include <QDebug>
 #include <QStandardPaths>
+#include <signal.h>
+#if KWINDOWSYSTEM_HAVE_X11
 #include <X11/Xlib.h>
 #include <fixx11h.h>
 #include <kwindowsystem.h>
 #include <kx11extras.h>
 #include <kxmessages.h>
-#include <private/qtx11extras_p.h>
-#include <signal.h>
+#endif
 
+#if KWINDOWSYSTEM_HAVE_X11
 static const char NET_STARTUP_MSG[] = "_NET_STARTUP_INFO";
-
+#endif
 // DESKTOP_STARTUP_ID is used also in kinit/wrapper.c ,
 // kdesu in both kdelibs and kdebase and who knows where else
 static const char NET_STARTUP_ENV[] = "DESKTOP_STARTUP_ID";
@@ -149,14 +162,18 @@ public:
     QMap<KStartupInfoId, KStartupInfo::Data> silent_startups;
     // contains ASN's that had change: but no new: yet
     QMap<KStartupInfoId, KStartupInfo::Data> uninited_startups;
+#if KWINDOWSYSTEM_HAVE_X11
     KXMessages msgs;
+#endif
     QTimer *cleanup;
     int flags;
 
     Private(int flags_P, KStartupInfo *qq)
         : q(qq)
         , timeout(60)
+#if KWINDOWSYSTEM_HAVE_X11
         , msgs(NET_STARTUP_MSG)
+#endif
         , cleanup(nullptr)
         , flags(flags_P)
     {
@@ -164,6 +181,7 @@ public:
 
     void createConnections()
     {
+#if KWINDOWSYSTEM_HAVE_X11
         // d == nullptr means "disabled"
         if (!QX11Info::isPlatformX11() || !QX11Info::display()) {
             return;
@@ -175,6 +193,7 @@ public:
         QObject::connect(&msgs, SIGNAL(gotMessage(QString)), q, SLOT(got_message(QString)));
         cleanup = new QTimer(q);
         QObject::connect(cleanup, SIGNAL(timeout()), q, SLOT(startups_cleanup()));
+#endif
     }
 };
 
@@ -192,6 +211,7 @@ KStartupInfo::~KStartupInfo()
 
 void KStartupInfo::Private::got_message(const QString &msg_P)
 {
+#if KWINDOWSYSTEM_HAVE_X11
     // TODO do something with SCREEN= ?
     // qCDebug(LOG_KWINDOWSYSTEM) << "got:" << msg_P;
     QString msg = msg_P.trimmed();
@@ -202,6 +222,9 @@ void KStartupInfo::Private::got_message(const QString &msg_P)
     } else if (msg.startsWith(QLatin1String("remove:"))) { // must match length below
         got_remove_startup_info(msg.mid(7));
     }
+#else
+    Q_UNUSED(msg_P)
+#endif
 }
 
 // if the application stops responding for a while, KWindowSystem may get
@@ -220,7 +243,11 @@ public:
         , w(w_P)
     {
     }
+#if KWINDOWSYSTEM_HAVE_X11
     Window w;
+#else
+    WId w;
+#endif
     static Type uniqueType()
     {
         return Type(QEvent::User + 15);
@@ -235,9 +262,11 @@ void KStartupInfo::Private::slot_window_added(WId w_P)
 
 void KStartupInfo::customEvent(QEvent *e_P)
 {
+#if KWINDOWSYSTEM_HAVE_X11
     if (e_P->type() == DelayedWindowEvent::uniqueType()) {
         d->window_added(static_cast<DelayedWindowEvent *>(e_P)->w);
     } else
+#endif
         QObject::customEvent(e_P);
 }
 
@@ -414,7 +443,12 @@ bool KStartupInfo::sendStartup(const KStartupInfoId &id_P, const KStartupInfoDat
     if (id_P.isNull()) {
         return false;
     }
+#if KWINDOWSYSTEM_HAVE_X11
     return sendStartupXcb(QX11Info::connection(), QX11Info::appScreen(), id_P, data_P);
+#else
+    Q_UNUSED(data_P)
+#endif
+    return true;
 }
 
 bool KStartupInfo::sendStartupXcb(xcb_connection_t *conn, int screen, const KStartupInfoId &id_P, const KStartupInfoData &data_P)
@@ -422,12 +456,19 @@ bool KStartupInfo::sendStartupXcb(xcb_connection_t *conn, int screen, const KSta
     if (id_P.isNull()) {
         return false;
     }
+#if KWINDOWSYSTEM_HAVE_X11
     QString msg = QStringLiteral("new: %1 %2").arg(id_P.d->to_text(), data_P.d->to_text());
     msg = Private::check_required_startup_fields(msg, data_P, screen);
 #ifdef KSTARTUPINFO_ALL_DEBUG
     qCDebug(LOG_KWINDOWSYSTEM) << "sending " << msg;
 #endif
     return KXMessages::broadcastMessageX(conn, NET_STARTUP_MSG, msg, screen);
+#else
+    Q_UNUSED(conn)
+    Q_UNUSED(screen)
+    Q_UNUSED(data_P)
+    return true;
+#endif
 }
 
 QString KStartupInfo::Private::check_required_startup_fields(const QString &msg, const KStartupInfoData &data_P, int screen)
@@ -452,7 +493,12 @@ bool KStartupInfo::sendChange(const KStartupInfoId &id_P, const KStartupInfoData
     if (id_P.isNull()) {
         return false;
     }
+#if KWINDOWSYSTEM_HAVE_X11
     return sendChangeXcb(QX11Info::connection(), QX11Info::appScreen(), id_P, data_P);
+#else
+    Q_UNUSED(data_P)
+#endif
+    return true;
 }
 
 bool KStartupInfo::sendChangeXcb(xcb_connection_t *conn, int screen, const KStartupInfoId &id_P, const KStartupInfoData &data_P)
@@ -460,11 +506,18 @@ bool KStartupInfo::sendChangeXcb(xcb_connection_t *conn, int screen, const KStar
     if (id_P.isNull()) {
         return false;
     }
+#if KWINDOWSYSTEM_HAVE_X11
     QString msg = QStringLiteral("change: %1 %2").arg(id_P.d->to_text(), data_P.d->to_text());
 #ifdef KSTARTUPINFO_ALL_DEBUG
     qCDebug(LOG_KWINDOWSYSTEM) << "sending " << msg;
 #endif
     return KXMessages::broadcastMessageX(conn, NET_STARTUP_MSG, msg, screen);
+#else
+    Q_UNUSED(conn)
+    Q_UNUSED(screen)
+    Q_UNUSED(data_P)
+    return true;
+#endif
 }
 
 bool KStartupInfo::sendFinish(const KStartupInfoId &id_P)
@@ -472,7 +525,10 @@ bool KStartupInfo::sendFinish(const KStartupInfoId &id_P)
     if (id_P.isNull()) {
         return false;
     }
+#if KWINDOWSYSTEM_HAVE_X11
     return sendFinishXcb(QX11Info::connection(), QX11Info::appScreen(), id_P);
+#endif
+    return true;
 }
 
 bool KStartupInfo::sendFinishXcb(xcb_connection_t *conn, int screen, const KStartupInfoId &id_P)
@@ -480,38 +536,60 @@ bool KStartupInfo::sendFinishXcb(xcb_connection_t *conn, int screen, const KStar
     if (id_P.isNull()) {
         return false;
     }
+#if KWINDOWSYSTEM_HAVE_X11
     QString msg = QStringLiteral("remove: %1").arg(id_P.d->to_text());
 #ifdef KSTARTUPINFO_ALL_DEBUG
     qCDebug(LOG_KWINDOWSYSTEM) << "sending " << msg;
 #endif
     return KXMessages::broadcastMessageX(conn, NET_STARTUP_MSG, msg, screen);
+#else
+    Q_UNUSED(conn)
+    Q_UNUSED(screen)
+    return true;
+#endif
 }
 
 bool KStartupInfo::sendFinish(const KStartupInfoId &id_P, const KStartupInfoData &data_P)
 {
-    //    if( id_P.isNull()) // id may be null, the pids and hostname matter then
-    //        return false;
+//    if( id_P.isNull()) // id may be null, the pids and hostname matter then
+//        return false;
+#if KWINDOWSYSTEM_HAVE_X11
     return sendFinishXcb(QX11Info::connection(), QX11Info::appScreen(), id_P, data_P);
+#else
+    Q_UNUSED(id_P)
+    Q_UNUSED(data_P)
+#endif
+    return true;
 }
 
 bool KStartupInfo::sendFinishXcb(xcb_connection_t *conn, int screen, const KStartupInfoId &id_P, const KStartupInfoData &data_P)
 {
-    //    if( id_P.isNull()) // id may be null, the pids and hostname matter then
-    //        return false;
+//    if( id_P.isNull()) // id may be null, the pids and hostname matter then
+//        return false;
+#if KWINDOWSYSTEM_HAVE_X11
     QString msg = QStringLiteral("remove: %1 %2").arg(id_P.d->to_text(), data_P.d->to_text());
 #ifdef KSTARTUPINFO_ALL_DEBUG
     qCDebug(LOG_KWINDOWSYSTEM) << "sending " << msg;
 #endif
     return KXMessages::broadcastMessageX(conn, NET_STARTUP_MSG, msg, screen);
+#else
+    Q_UNUSED(conn)
+    Q_UNUSED(screen)
+    Q_UNUSED(id_P)
+    Q_UNUSED(data_P)
+    return true;
+#endif
 }
 
 void KStartupInfo::appStarted()
 {
     QByteArray startupId = s_startup_id;
 
+#if KWINDOWSYSTEM_HAVE_X11
     if (startupId.isEmpty()) {
         startupId = QX11Info::nextStartupId();
     }
+#endif
 
     appStarted(startupId);
     setStartupId("0"); // reset the id, no longer valid (must use clearStartupId() to avoid infinite loop)
@@ -524,9 +602,11 @@ void KStartupInfo::appStarted(const QByteArray &startup_id)
     if (id.isNull()) {
         return;
     }
+#if KWINDOWSYSTEM_HAVE_X11
     if (QX11Info::isPlatformX11() && !qEnvironmentVariableIsEmpty("DISPLAY")) { // don't rely on QX11Info::display()
         KStartupInfo::sendFinish(id);
     }
+#endif
 }
 
 void KStartupInfo::setStartupId(const QByteArray &startup_id)
@@ -538,6 +618,7 @@ void KStartupInfo::setStartupId(const QByteArray &startup_id)
         s_startup_id = "0";
     } else {
         s_startup_id = startup_id;
+#if KWINDOWSYSTEM_HAVE_X11
         if (QX11Info::isPlatformX11()) {
             KStartupInfoId id;
             id.initId(startup_id);
@@ -551,6 +632,7 @@ void KStartupInfo::setStartupId(const QByteArray &startup_id)
                 }
             }
         }
+#endif
     }
 }
 
@@ -558,6 +640,7 @@ void KStartupInfo::setNewStartupId(QWindow *window, const QByteArray &startup_id
 {
     Q_ASSERT(window);
     setStartupId(startup_id);
+#if KWINDOWSYSTEM_HAVE_X11
     bool activate = true;
     if (window != nullptr && QX11Info::isPlatformX11()) {
         if (!startup_id.isEmpty() && startup_id != "0") {
@@ -576,6 +659,9 @@ void KStartupInfo::setNewStartupId(QWindow *window, const QByteArray &startup_id
             KX11Extras::forceActiveWindow(window->winId());
         }
     }
+#else
+    Q_UNUSED(window)
+#endif
 }
 
 KStartupInfo::startup_t KStartupInfo::checkStartup(WId w_P, KStartupInfoId &id_O, KStartupInfoData &data_O)
@@ -619,6 +705,7 @@ KStartupInfo::startup_t KStartupInfo::Private::check_startup_internal(WId w_P, K
         }
         return find_id(id, id_O, data_O) ? Match : NoMatch;
     }
+#if KWINDOWSYSTEM_HAVE_X11
     if (!QX11Info::isPlatformX11()) {
         qCDebug(LOG_KWINDOWSYSTEM) << "check_startup:cantdetect";
         return CantDetect;
@@ -652,6 +739,7 @@ KStartupInfo::startup_t KStartupInfo::Private::check_startup_internal(WId w_P, K
     if (transient_for != QX11Info::appRootWindow() && transient_for != XCB_WINDOW_NONE) {
         return NoMatch;
     }
+#endif
     qCDebug(LOG_KWINDOWSYSTEM) << "check_startup:cantdetect";
     return CantDetect;
 }
@@ -721,6 +809,7 @@ bool KStartupInfo::Private::find_wclass(const QByteArray &_res_name, const QByte
 
 QByteArray KStartupInfo::windowStartupId(WId w_P)
 {
+#if KWINDOWSYSTEM_HAVE_X11
     if (!QX11Info::isPlatformX11()) {
         return QByteArray();
     }
@@ -732,10 +821,15 @@ QByteArray KStartupInfo::windowStartupId(WId w_P)
         ret = groupLeaderInfo.startupId();
     }
     return ret;
+#else
+    Q_UNUSED(w_P)
+    return QByteArray();
+#endif
 }
 
 void KStartupInfo::Private::setWindowStartupId(WId w_P, const QByteArray &id_P)
 {
+#if KWINDOWSYSTEM_HAVE_X11
     if (!QX11Info::isPlatformX11()) {
         return;
     }
@@ -744,6 +838,10 @@ void KStartupInfo::Private::setWindowStartupId(WId w_P, const QByteArray &id_P)
     }
     NETWinInfo info(QX11Info::connection(), w_P, QX11Info::appRootWindow(), NET::Properties(), NET::Properties2());
     info.setStartupId(id_P.constData());
+#else
+    Q_UNUSED(w_P)
+    Q_UNUSED(id_P)
+#endif
 }
 
 void KStartupInfo::setTimeout(unsigned int secs_P)
@@ -813,9 +911,11 @@ void KStartupInfo::Private::clean_all_noncompliant()
 QByteArray KStartupInfo::createNewStartupId()
 {
     quint32 timestamp = 0;
+#if KWINDOWSYSTEM_HAVE_X11
     if (QX11Info::isPlatformX11()) {
         timestamp = QX11Info::getTimestamp();
     }
+#endif
     return KStartupInfo::createNewStartupIdForTimestamp(timestamp);
 }
 
@@ -824,7 +924,15 @@ QByteArray KStartupInfo::createNewStartupIdForTimestamp(quint32 timestamp)
     // Assign a unique id, use hostname+time+pid, that should be 200% unique.
     // Also append the user timestamp (for focus stealing prevention).
     struct timeval tm;
+#ifdef Q_OS_WIN
+    // on windows only msecs accuracy instead of usecs like with gettimeofday
+    // XXX: use Win API to get better accuracy
+    qint64 msecsSinceEpoch = QDateTime::currentMSecsSinceEpoch();
+    tm.tv_sec = msecsSinceEpoch / 1000;
+    tm.tv_usec = (msecsSinceEpoch % 1000) * 1000;
+#else
     gettimeofday(&tm, nullptr);
+#endif
     char hostname[256];
     hostname[0] = '\0';
     if (!gethostname(hostname, 255)) {
diff --git a/src/kstartupinfo.h b/src/kstartupinfo.h
index 22256dd..548dc0e 100644
--- a/src/kstartupinfo.h
+++ b/src/kstartupinfo.h
@@ -21,6 +21,10 @@ typedef struct _XDisplay Display;
 
 struct xcb_connection_t;
 
+#ifdef Q_CC_MSVC
+typedef int pid_t;
+#endif
+
 class KStartupInfoId;
 class KStartupInfoData;
 
diff --git a/src/kwindowsystem.cpp b/src/kwindowsystem.cpp
index 87cc4c6..0798105 100644
--- a/src/kwindowsystem.cpp
+++ b/src/kwindowsystem.cpp
@@ -5,15 +5,12 @@
 */
 #include "kwindowsystem.h"
 #include "kwindowsystem_debug.h"
+#include "kstartupinfo.h"
 #include "kwindowsystem_dummy_p.h"
 #include "pluginwrapper_p.h"
 
 #include <config-kwindowsystem.h>
 
-#if KWINDOWSYSTEM_HAVE_X11
-#include "kstartupinfo.h"
-#endif
-
 #include <QGuiApplication>
 #include <QMetaMethod>
 #include <QPixmap>
